kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java

```diff
@@ -54,44 +55,66 @@
+    /**
+     * Inner class defined to track the CloudServices as they get added, modified or removed.
+     * Specific methods can refresh the cloudService definition and setup again the Cloud Client.
+     *
+     */
+    private final class CloudPublisherServiceTrackerCustomizer
+            implements ServiceTrackerCustomizer<CloudService, CloudService> {
+
+        @Override
+        public CloudService addingService(ServiceReference<CloudService> reference) {
+            CloudPublisher.this.cloudService = CloudPublisher.this.componentContext.getBundleContext()
+                    .getService(reference);
+            try {
+                // recreate the Cloud Client
+                setupCloudClient();
+            } catch (KuraException e) {
+                s_logger.error(s_message.cloudClientSetupProblem() + ThrowableUtil.stackTraceAsString(e));
+            }
+            return CloudPublisher.this.cloudService;
+        }
+
+        @Override
+        public void modifiedService(ServiceReference<CloudService> reference, CloudService service) {
+            CloudPublisher.this.cloudService = CloudPublisher.this.componentContext.getBundleContext()
+                    .getService(reference);
+            try {
+                // recreate the Cloud Client
+                setupCloudClient();
+            } catch (KuraException e) {
+                s_logger.error(s_message.cloudClientSetupProblem() + ThrowableUtil.stackTraceAsString(e));
+            }
+        }
+
+        @Override
+        public void removedService(ServiceReference<CloudService> reference, CloudService service) {
+            CloudPublisher.this.cloudService = null;
+        }
+    }
 
-    /** The Logger instance. */
     private static final Logger s_logger = LoggerFactory.getLogger(CloudPublisher.class);
 
-    /** Localization Resource */
     private static final WireMessages s_message = LocalizationAdapter.adapt(WireMessages.class);
 
-    /** The cloud client. */
     private CloudClient cloudClient;
 
-    /** The cloud service. */
     private volatile CloudService cloudService;
 
-    /** The data service. */
-    private volatile DataService dataService;
+    private ServiceTracker<CloudService, CloudService> cloudServiceTracker;
 
-    /** Synchronization Monitor. */
-    private final Lock monitor;
-
-    /** The cloud publisher options. */
     private CloudPublisherOptions options;
 
-    /** The Wire Helper Service. */
     private volatile WireHelperService wireHelperService;
 
-    /** The wire supporter component. */
     private WireSupport wireSupport;
 
-    /**
-     * Instantiates a new cloud publisher instance.
-     */
-    public CloudPublisher() {
-        this.monitor = new ReentrantLock();
-    }
+    private ComponentContext componentContext;
+
+    private final ServiceTrackerCustomizer<CloudService, CloudService> customizer = new CloudPublisherServiceTrackerCustomizer();
```

Comment:
The tracker was already instantiated in the activate method.
The trackerCustomizer, used by the tracker was initialized at the class instantiation. I've moved that into the activate method as is close to the initialization of the tracker, even if I don't believe we will save so much...